Алешев Руслан

artBETEP opened the pull request:
Реализованы базовые требования.
Сделана сохранение изображений при ошибке.
Добавлен README.md и изображения с примерами.

Commit:
Prechecking

Реализованы базовые требования.
Сделана сохранение изображений при ошибке.
Добавлен README.md и изображения с примерами.

artBETEP committed:
Update README.md

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+    public class CircularCloudLayouter
+    {
+        private readonly Point center;
+        private List<Rectangle> cloud;
```
Лучше `IList`, а то и вовсе `ICollection`. Это позволит не привязываться к реализации коллекции и поменять ее в любой момент

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+        private List<Rectangle> cloud;
+
+        //Spiral coeffs
+        int i = 0;
+        float it = (float)Math.PI / 21;
+        float ri = 50;
```
Нужны говорящие имена, модификаторы доступа, мб еще какие-нибудь. Можно даже `summary`, поясняющее физический смысл полей

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            if (center.X <= 0 || center.Y <= 0)
+                throw new ArgumentException("Central point coordinates should be in positive");
+            cloud = new List<Rectangle>();
```
В `CreateCloud` всё равно будет создан новый лист. Мб тогда здесь это не нужно

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+
+            while (true)
+            {
+                bool findPlace = true;
```
Здесь и во многих других местах - `var`

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+            return rect;
+        }
+
+        public List<Rectangle> CreateCloud(List<Size> rectangleSizes)
```
Тут 2 варианта:
- Либо менять возвращаемый тип на `IReadonlyCollection`, чтобы никто снаружи при получении списка прямоугольников не мог на него как-то влиять (очистить, например)
- Либо не возвращать здесь вообще ничего, завести отдельную проперть и использовать ее во всех необходимых местах

Еще, мне как потенциальному пользователю твоего класса не ясна его роль. Например, что если я вызову этот метод 2 раза? Список старых прямоугольников очистится, но функция продолжить работать со старого индекса, получится не оч. Нужно решить как оно должно себя вести - пересоздавать или дополнять и дать соответствующее название

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+            return cloud;
+        }
+
+        public Image CreateImage()
```
Обычно такое называют `To[что-нибудь]`. Как `ToArray` ниже, например. Мб `ToImage`?

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+            return image;
+        }
+
+        private Point GetNextPoint()
```
Если знаком с перечислениями, было бы наглядно переписать на `IEnumerable<Point>` с `yield return`. Таким образом получится спрятать неприятный бесконечный цикл в этой небольшой функции

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            cloud = new List<Rectangle>();
+
+            foreach (var rectangleSize in rectangleSizes)
```
Если знаком с `LINQ`, можно схлопнуть всю функцию буквально в одну строку

a-zhelonkin commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,99 @@
+                var point = GetNextPoint();
+
+                rect = new Rectangle(new Point(point.X - rectSize.Width / 2, point.Y - rectSize.Height / 2), rectSize);
+                foreach (var previous in cloud)
+                {
+                    if (rect.IntersectsWith(previous))
+                    {
+                        findPlace = false;
+                        break;
+                    }
+                }
```
Если знаком с `LINQ`, можно схлопнуть всю функцию буквально в одну строку

a-zhelonkin commented on cs/TagsCloudVisualization/TagCloudVisualisationTest.cs:
```diff
@@ -0,0 +1,122 @@
+        [TestCase(123)]
+        public void CreateCloud_ReturnCorrectNumberOfRectangles(int rectCount)
+        {
+            var layouter = new CircularCloudLayouter(new Point(50, 50));
```
- Давай явно обозначим комментариями где какая часть AAA
- Объект тестируемого класса часто называются `sut` (system under testing). Это помогает быстрее ориентироваться в тесте

a-zhelonkin commented on cs/TagsCloudVisualization/TagCloudVisualisationTest.cs:
```diff
@@ -0,0 +1,122 @@
+        [TestCase(10)]
+        [TestCase(20)]
+        [TestCase(200)]
+        public void CreateCloud_RectanglesShouldNotIntersect(int rectCount)
```
Если это попытка отрисовать облако при падении теста, то кажется что ты несколько неправильно понял задачу. Тебе нужен некий кусок кода, который вызовется при невыполнении любого условия из тестов выше, откуда-то возьмет входные данные теста и попробует отрисовать

Commit:
Refactoring

Создание точек вынесено в отдельный класс с интерфейсом IPointsProvider
При падении тестов на раскладку сохраняется файл с изображением
Поправлены замечания

Commit:
Merge branch 'master' of https://github.com/artBETEP/tdd_homework
